### PR TITLE
move event witnesses directly into the participant_events table

### PR DESCRIPTION
**Pull request by leo-da**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableFlatEvents.scala [hunk 1]
```diff
@@ -74,10 +75,10 @@ private[events] trait EventsTableFlatEvents { this: EventsTable =>
     ).mkString(", ")
 
   private val witnessesAggregation =
-    "array_agg(event_witness) as event_witnesses"
+    "flat_event_witnesses"
 
   private val flatEventsTable =
-    "participant_events natural join participant_event_flat_transaction_witnesses"
+    "participant_events"
```
Comment (stefanobaghino-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableFlatEvents.scala:81]: Feel free to inline them in the actual queries.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableQueries.scala [hunk 2]
```diff
@@ -22,4 +22,18 @@ object EventsTableQueries {
     lastEventNodeIndexFromPreviousPage
       .map(x => (lastOffsetFromPreviousPage, x))
       .getOrElse(NonExistingOffsetNodeIndexPair)
+
+  def format(ps: Set[Party]): String =
+    ps.view.map(p => s"'$p'").mkString(",")
```
Comment (stefanobaghino-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableQueries.scala:27]: This formatting should already be applied by Anorm and should cache queries when possible, any reason to not use that?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableTreeEvents.scala [hunk 3]
```diff
@@ -86,10 +89,10 @@ private[events] trait EventsTableTreeEvents { this: EventsTable =>
   ).mkString(", ")
 
   private val witnessesAggregation =
-    "array_agg(event_witness) as event_witnesses"
+    "tree_event_witnesses"
 
   private val treeEventsTable =
-    "participant_events natural join participant_event_transaction_tree_witnesses"
+    "participant_events"
```
Comment (stefanobaghino-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableTreeEvents.scala:95]: These can also be folded in the query itself at this point. They were meant to be short-hands for longer fragments.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableInsert.scala [hunk 4]
```diff
@@ -160,13 +158,15 @@ private[events] trait EventsTableInsert { this: EventsTable =>
     * @throws RuntimeException If a value cannot be serialized into an array of bytes
     */
   @throws[RuntimeException]
-  def prepareBatchInsert(
+  def prepareBatchInsert[Nid <: NodeId](
```
Comment (stefanobaghino-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableInsert.scala:161]: `NodeId` is a `final case class`, this is probably unnecessarily generic.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableQueries.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableQueries.scala:27]: I just saw that you use this to specify the array, marking as resolved.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableInsert.scala (hunk 4 above)
Comment (S11001001) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableInsert.scala:161]: Scala seems to think that `Nid` and `NodeId` are distinct, which is [exploitable in type-checking](https://typelevel.org/blog/2017/02/13/more-types-than-classes.html).

Comment (S11001001) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableInsert.scala:161]: To expand, there is no need for separate maps here at all, I would have rather mapped the party sets directly into the transaction structure. However, the transaction structure isn't sufficiently generic for that, so we have the nasty Perlish technique of "pass a map that associates the extra data in" instead.

Then I ask myself, "given that I don't want to hypergeneralize lf.Transaction, how can I make clear that these three arguments are related in a specific way, and they do not merely contain NodeIds, but NodeIds that are specifically related to each other?" And this is the most obvious tool available.

[I expand on a similar case for similarly final class String.](https://typelevel.org/blog/2017/09/05/three-types-of-strings.html)

**Pull request merged**